### PR TITLE
feat(mission_planner): add mrm reroute implementation

### PR DESCRIPTION
**Pull request by purewater0901**
## Description
Add implementation for mrm reroute

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 1]
```diff
@@ -273,19 +302,112 @@ void MissionPlanner::on_set_mrm_route(
...
+  points.push_back(odometry_->pose.pose);
+  points.push_back(normal_route_->goal_pose);
+
+  // Plan route.
+  LaneletRoute new_route = planner_->plan(points);
+  if (new_route.segments.empty()) {
+    RCLCPP_ERROR_THROTTLE(get_logger(), *get_clock(), 5000, "Reroute with normal goal failed.");
+    res->success = false;
+    return;
+  }
```
Comment (kosuke55) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:401]: I think it also is better to check `checkRerouteSafety(*mrm_route_, new_route)`.
if it is `false` return

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 2]
```diff
@@ -273,19 +302,112 @@ void MissionPlanner::on_set_mrm_route(
...
+  new_route.header.stamp = odometry_->header.stamp;
+  new_route.header.frame_id = map_frame_;
+  new_route.uuid.uuid = generate_random_id();
+  new_route.allow_modification = true;
```
Comment (kosuke55) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:405]:
```suggestion
  new_route.allow_modification = normal_route_->allow_modification;
```

File: planning/mission_planner/src/mission_planner/mission_planner.cpp (hunk 1 above)
Comment (kosuke55) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:401]: and change_route(*mrm_route_)

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 3]
```diff
@@ -297,19 +419,97 @@ void MissionPlanner::on_modified_goal(const ModifiedGoal::Message::ConstSharedPt
...
-  // TODO(Yutaka Shimizu): reroute
-  (void)req;
-  (void)res;
+  using ResponseCode = autoware_adapi_v1_msgs::srv::SetRoute::Response;
+
+  if (!odometry_) {
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "The vehicle pose is not received.");
+  }
+
```
Comment (kosuke55) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:428]: block setting normal route if emergency

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 4]
```diff
@@ -297,19 +419,97 @@ void MissionPlanner::on_modified_goal(const ModifiedGoal::Message::ConstSharedPt
...
+  if (!planner_->ready()) {
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "The planner is not ready.");
+  }
+  if (!odometry_) {
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "The vehicle pose is not received.");
+  }
+
```
Comment (kosuke55) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:475]: block setting normal route if emergency

Comment (kosuke55) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:475]: and set_route/route_points

Comment (kosuke55) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:475]: "emergency" means that mrm_route is used

File: planning/mission_planner/src/mission_planner/mission_planner.cpp (hunk 3 above)
Comment (kosuke55) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:428]: RouteState::Message::CHANGING

File: planning/mission_planner/src/mission_planner/mission_planner.cpp (hunk 4 above)
Comment (kosuke55) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:475]: RouteState::Message::CHANGING

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 5]
```diff
@@ -161,6 +169,25 @@ void MissionPlanner::change_route(const LaneletRoute & route)
   pub_route_->publish(route);
   pub_marker_->publish(planner_->visualize(route));
   planner_->updateRoute(route);
+
+  // update normal route
+  normal_route_ = std::make_shared<LaneletRoute>(route);
+}
+
+void MissionPlanner::change_mrm_route(const LaneletRoute & route)
+{
```
Comment (kosuke55) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:178]: RouteState::Message::CHANGING

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 6]
```diff
@@ -273,19 +302,112 @@ void MissionPlanner::on_set_mrm_route(
...
+    res->status.success = false;
+  }
 }
 
 // NOTE: The route interface should be mutually exclusive by callback group.
 void MissionPlanner::on_clear_mrm_route(
-  const ClearMrmRoute::Service::Request::SharedPtr req,
+  [[maybe_unused]] const ClearMrmRoute::Service::Request::SharedPtr req,
   const ClearMrmRoute::Service::Response::SharedPtr res)
 {
```
Comment (kosuke55) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:362]: RouteState::Message::CHANGING

File: planning/mission_planner/src/mission_planner/mission_planner.cpp (hunk 4 above)
Comment (kosuke55) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:475]: e.g. set is_emegency_ true if on_mrm_route is success

File: planning/mission_planner/src/mission_planner/mission_planner.cpp (hunk 1 above)
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp:402]: done https://github.com/autowarefoundation/autoware.universe/pull/3466/commits/dc175782d2141c241ee9061b23f81ee75de3f66c

File: planning/mission_planner/src/mission_planner/mission_planner.cpp (hunk 2 above)
Comment (purewater0901) [reply planning/mission_planner/src/mission_planner/mission_planner.cpp]: done https://github.com/autowarefoundation/autoware.universe/pull/3466/commits/dc175782d2141c241ee9061b23f81ee75de3f66c

**Pull request closed without merging**